src/util/simplify_expr.cpp

```diff
@@ -158,6 +161,78 @@ bool simplify_exprt::simplify_popcount(popcount_exprt &expr)
   return true;
 }
 
+bool simplify_exprt::simplify_function_application(exprt &rhs)
+{
+  const function_application_exprt &function_app =
+    to_function_application_expr(rhs);
+  irep_idt func_id =
+    function_app.function().find(ID_expression).get(ID_identifier);
+
+  // Starts-with is used for .equals.
+  if(func_id == ID_cprover_string_startswith_func)
+  {
+    // We want to get both arguments of any starts-with comparison, and
+    // trace them back to the actual string instance. All variables on the
+    // way must be constant for us to be sure this will work.
+    auto &first_argument = to_string_expr(function_app.arguments().at(0));
+    auto &second_argument = to_string_expr(function_app.arguments().at(1));
+
+    if(
+      first_argument.content().id() != ID_address_of ||
+      second_argument.content().id() != ID_address_of)
+      return true;
+
+    const address_of_exprt &first_address_of =
+      to_address_of_expr(first_argument.content());
+    const address_of_exprt &second_address_of =
+      to_address_of_expr(second_argument.content());
+
+    // Constants are got via address_of expressions, so this helps filter out
+    // things that are non-constant.
+    if(
+      first_address_of.object().id() != ID_index ||
+      first_address_of.object().id() != ID_index)
+      return true;
+
+    const index_exprt &first_index_expression =
+      to_index_expr(first_address_of.object());
+    const index_exprt &second_index_expression =
+      to_index_expr(second_address_of.object());
+
+    const exprt &first_string_array = first_index_expression.array();
+    const exprt &second_string_array = second_index_expression.array();
+
+    // Check if our symbols type is an array.
+    if(
+      first_string_array.type().id() != ID_array ||
+      second_string_array.type().id() != ID_array)
+      return true;
+
+    // If so, it'll be a string array that we can use.
+    const array_string_exprt &first_string = to_array_string_expr(
+      this->ns.lookup(first_string_array.get(ID_identifier)).value);
+
+    const array_string_exprt &second_string = to_array_string_expr(
+      this->ns.lookup(second_string_array.get(ID_identifier)).value);
+
+    // If the lengths are the same we can do the .equals substitute.
+    if(first_string.length() == second_string.length())
+    {
+      exprt result;
+      if(first_string == second_string)
+        result = from_integer(1, c_bool_typet(8));
+      else
+        result = from_integer(0, c_bool_typet(8));
```

Comment:
I'd do:
`expr = from_integer(first_string == second_string ? 1 : 0, expr.type());`
It's shorter, but most importantly it ensures that you preserve the type.